### ISSUE TITLE
Sync users with unix-users?

### ISSUE THREAD

**Issue by nreith** [jhuckaby/Cronicle#373]
Does anyone know how I can set cronicle to use the same usernames and passwords as users on a server?

**Comment by mikeTWC1984** [comment]
Nah, out of the box  Cronicle can only use it's own auth. You can also set up external auth, but it should be configured to mirror Cronicle auth api (like store permission metadata same way), which is not trivial. I managed to tweak internal auth to use external credential verification (AD/LDAP) instead of local hash checking. This might be adopted to use local unix creds, but I don't think it's very useful for local users - you basically need to set up user 2 times. I also couldn't find any node module for local cred checking. 
You may allow open accounts, so user can register themselves using same creds.

**Comment by nreith** [comment]
I see. I may go the LDAP/AD route. Any tips on getting that set up?
Thanks for the reply btw!

**Comment by mikeTWC1984** [comment]
This would require some source code update. I'll share some code later

**Comment by mikeTWC1984** [comment]
Here is the updated user module I created for my fork to enable AD auth
https://github.com/cronicle-edge/cronicle-edge/blob/main/lib/user.js
In order to use it within existing cronicle:
 - copy this user.js file to your lib folder
 - install activedirectory2 module (npm i activedirectory2)
 - add ad_domain config (should refer to your ad domain, smth like corp.yourdomain.com)
 - update lib/main.js to use new module instead of pixl-user ( replace *require('pixl-server-user')* with *require('./user.js')* )

The new module basically adds few [extra lines](https://github.com/cronicle-edge/cronicle-edge/blob/main/lib/user.js#L207-L220)  to alter password verification if user has ext_auth flag checked. In this case its using AD verify function, but you can replace it with anything else that can verify username/password (e.g. some post request)

Now if you need to add user with external auth you need to check that ext_auth flag. It's kind of tedious to implement it on UI. I guess the easiest way to set it up on existing cronicle is to extend bin/storage-cli.js script. It already can add admin user, so you can add similar option for ad users. Add below snippet after "case admin" option
```js
		case 'aduser':
			// Usage: ./storage-cli.js aduser USERNAME FULLNAME [EMAIL]
			var username = commands.shift();
			var fullname = commands.shift();
			// dummy password
			var password = Tools.generateUniqueID(32, username);
			var email = commands.shift() || username + '@cronicle.com';
			if (!username || !password) {
				print("\nUsage: bin/storage-cli.js admin USERNAME PASSWORD [EMAIL]\n\n");
				process.exit(1);
			}
			if (!username.match(/^[\w\.\-]+@?[\w\.\-]+$/)) {
				print("\nERROR: Username must contain only alphanumerics, dash and period.\n\n");
				process.exit(1);
			}
			username = username.toLowerCase();

			var user = {
				username: username,
				password: password,
				full_name: fullname || username + ' (AD)',
				email: email,
				ext_auth: true,
			};

			user.active = 1;
			user.created = user.modified = Tools.timeNow(true);
			user.salt = Tools.generateUniqueID(64, user.username);
			user.password = bcrypt.hashSync(user.password + user.salt);
			user.privileges = { run_events: 1 };

			storage.put('users/' + username, user, function (err) {
				if (err) throw err;
				storage.listUnshift('global/users', { username: user.username }, function (err) {
					if (err) self.logError(1, "Failed to add user to manager list: " + err);

					print("\nAD user '" + username + "' created successfully.\n");
					print("\n");
	
					storage.shutdown(function () { process.exit(0); });

				});

			});
			break;
```
Then you can add add user this way
```bash
bin/storage-cli.js aduser someuser "John Smith" jonh.smith@cronicle.com
```
Just make sure you stop cronicle when running this. Alternatively you can try out my fork. It's kind of experimental but most features are stable.

**Comment by nreith** [comment]
Thanks man!

I think for the moment I'm sticking with simple auth. But I may check this
out in the future.

**Comment by mouatez25** [comment]
> Here is the updated user module I created for my fork to enable AD auth https://github.com/cronicle-edge/cronicle-edge/blob/main/lib/user.js In order to use it within existing cronicle:
> 
> * copy this user.js file to your lib folder
> * install activedirectory2 module (npm i activedirectory2)
> * add ad_domain config (should refer to your ad domain, smth like corp.yourdomain.com)
> * update lib/main.js to use new module instead of pixl-user ( replace _require('pixl-server-user')_ with _require('./user.js')_ )
> 
> The new module basically adds few [extra lines](https://github.com/cronicle-edge/cronicle-edge/blob/main/lib/user.js#L207-L220) to alter password verification if user has ext_auth flag checked. In this case its using AD verify function, but you can replace it with anything else that can verify username/password (e.g. some post request)
> 
> Now if you need to add user with external auth you need to check that ext_auth flag. It's kind of tedious to implement it on UI. I guess the easiest way to set it up on existing cronicle is to extend bin/storage-cli.js script. It already can add admin user, so you can add similar option for ad users. Add below snippet after "case admin" option
> 
> ```js
> 		case 'aduser':
> 			// Usage: ./storage-cli.js aduser USERNAME FULLNAME [EMAIL]
> 			var username = commands.shift();
> 			var fullname = commands.shift();
> 			// dummy password
> 			var password = Tools.generateUniqueID(32, username);
> 			var email = commands.shift() || username + '@cronicle.com';
> 			if (!username || !password) {
> 				print("\nUsage: bin/storage-cli.js admin USERNAME PASSWORD [EMAIL]\n\n");
> 				process.exit(1);
> 			}
> 			if (!username.match(/^[\w\.\-]+@?[\w\.\-]+$/)) {
> 				print("\nERROR: Username must contain only alphanumerics, dash and period.\n\n");
> 				process.exit(1);
> 			}
> 			username = username.toLowerCase();
> 
> 			var user = {
> 				username: username,
> 				password: password,
> 				full_name: fullname || username + ' (AD)',
> 				email: email,
> 				ext_auth: true,
> 			};
> 
> 			user.active = 1;
> 			user.created = user.modified = Tools.timeNow(true);
> 			user.salt = Tools.generateUniqueID(64, user.username);
> 			user.password = bcrypt.hashSync(user.password + user.salt);
> 			user.privileges = { run_events: 1 };
> 
> 			storage.put('users/' + username, user, function (err) {
> 				if (err) throw err;
> 				storage.listUnshift('global/users', { username: user.username }, function (err) {
> 					if (err) self.logError(1, "Failed to add user to manager list: " + err);
> 
> 					print("\nAD user '" + username + "' created successfully.\n");
> 					print("\n");
> 	
> 					storage.shutdown(function () { process.exit(0); });
> 
> 				});
> 
> 			});
> 			break;
> ```
> 
> Then you can add add user this way
> 
> ```shell
> bin/storage-cli.js aduser someuser "John Smith" jonh.smith@cronicle.com
> ```
> 
> Just make sure you stop cronicle when running this. Alternatively you can try out my fork. It's kind of experimental but most features are stable.

thanks for sharing this solution , can you please provide example of 

- add ad_domain config (should refer to your ad domain, smth like corp.yourdomain.com)
